### PR TITLE
mgr/cephadm: stable nfs-ganesha ranks; nfs + ingress support

### PR DESCRIPTION
**Pull request by liewegas**
- Assign each ganesha daemon a stable rank id (0, 1, ...).
- Ensure that no more than one daemon at a time is running on a single rank.  For now, this relies on revoking the old daemon's auth key. In the future, we can additionally fence it, but that infrastructure isn't in tree yet.
- Add NFS support for ingress: haproxy with consistent hashing
- Update mgr/nfs to optionally deploy ingress (when VIP is provided)
- Update mgr/nfs cluster ls output to include stable endpoint separate from backend server instances
- Make everything default to pool ``nfs-ganesha`` and namespace matching the service_id
- Update mgr/nfs commands to use ``rm`` instead of ``delete`` for consistency with the rest of the CLI

### CODE REVIEW

**Comment by liewegas** [comment]
@jtlayton Is this the right thing? https://github.com/ceph/ceph/pull/41007/commits/49e15c762c71d5a58642f18eee835bbb38036ead#diff-b13a6b9b90f10d54533fcaa8b7d9db7f8d24ae39bb52cd611c28c9c39cb429ecR86

This goes into the template https://github.com/ceph/ceph/blob/master/src/pybind/mgr/cephadm/templates/services/nfs/ganesha.conf.j2

**Comment by jtlayton** [comment]
> @jtlayton Is this the right thing? [49e15c7#diff-b13a6b9b90f10d54533fcaa8b7d9db7f8d24ae39bb52cd611c28c9c39cb429ecR86](https://github.com/ceph/ceph/commit/49e15c762c71d5a58642f18eee835bbb38036ead#diff-b13a6b9b90f10d54533fcaa8b7d9db7f8d24ae39bb52cd611c28c9c39cb429ecR86)
> 
> This goes into the template https://github.com/ceph/ceph/blob/master/src/pybind/mgr/cephadm/templates/services/nfs/ganesha.conf.j2

Yeah, looks good. As long as you have unique nodeids for the daemons that are running, and their successors get the same name you should be fine.

**Review by mgfritch — COMMENTED** [review]
couple of suggestions, but looks good overall!

File: src/pybind/mgr/cephadm/inventory.py [hunk 1]
```diff
@@ -167,7 +167,7 @@ def load(self):
                 self._specs[service_name] = spec
                 self.spec_created[service_name] = created
 
-                if 'deleted' in v:
+                if 'deleted' in j:
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/inventory.py:173]: nice find!

File: src/pybind/mgr/orchestrator/_interface.py [hunk 2]
```diff
@@ -943,6 +943,7 @@ def to_json(self) -> dict:
         out: Dict[str, Any] = OrderedDict()
         out['daemon_type'] = self.daemon_type
         out['daemon_id'] = self.daemon_id
+        out['service_name'] = self._service_name
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/_interface.py:952]: I'd remove `osdspec_affinity` a few lines below by now.

File: src/pybind/mgr/cephadm/inventory.py [hunk 3]
```diff
@@ -171,14 +174,35 @@ def load(self):
                     deleted = str_to_datetime(cast(str, j['deleted']))
                     self.spec_deleted[service_name] = deleted
 
+                if 'rank_map' in j and isinstance(j['rank_map'], dict):
+                    self._rank_maps[service_name] = {}
+                    for rank_str, m in j['rank_map'].items():
+                        try:
+                            rank = int(rank_str)
+                        except ValueError:
+                            continue
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/inventory.py:183]:
```suggestion
                            logger.exception(f'failed to parse rank_str {j['rank_map']}')
                            continue
````

File: src/pybind/mgr/cephadm/inventory.py [hunk 4]
```diff
@@ -131,6 +132,7 @@ def __init__(self, mgr):
         # type: (CephadmOrchestrator) -> None
         self.mgr = mgr
         self._specs = {}  # type: Dict[str, ServiceSpec]
+        self._rank_maps = {}    # type: Dict[str, Dict[int, Dict[int, Optional[str]]]]
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/inventory.py:135]: Add a comment describing the map? Like so?
```
# service_name -> rank -> gen -> daemon_id
```

File: src/pybind/mgr/cephadm/schedule.py [hunk 5]
```diff
@@ -49,6 +49,52 @@ def renumber_ports(self, n: int) -> 'DaemonPlacement':
...
+        return DaemonPlacement(
+            self.daemon_type,
+            self.hostname,
+            self.network,
+            self.name,
+            self.ip,
+            self.ports,
+            rank,
+            gen,
+        )
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/schedule.py:62]: Is this the best place to put that code. Would be great to have the rank assignment in the `NFSService` class.

File: src/pybind/mgr/cephadm/schedule.py [hunk 6]
```diff
@@ -195,29 +265,40 @@ def expand_candidates(ls: List[DaemonPlacement], num: int) -> List[DaemonPlaceme
...
+        # consider (preserve) existing daemons in a particular order...
+        daemons = sorted(
+            [
+                d for d in self.daemons if d.daemon_type == self.primary_daemon_type
+            ],
+            key=lambda d: (
+                not d.is_active,               # active before standby
+                d.rank is not None,            # ranked first, then non-ranked
+                d.rank,                        # low ranks
+                0 - (d.rank_generation or 0),  # newer generations first
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/schedule.py:277]: Feels wrong to me. We're putting a lot of per-service logic into the scheduler here. Can we keep it generic in the scheduler?

File: src/pybind/mgr/cephadm/tests/test_scheduling.py [hunk 7]
```diff
@@ -379,19 +379,22 @@ class NodeAssignmentTest(NamedTuple):
...
     daemons: List[DaemonDescription]
+    rank_map: Optional[Dict[int, Dict[int, Optional[str]]]]
+    post_rank_map: Optional[Dict[int, Dict[int, Optional[str]]]]
     expected: List[str]
     expected_add: List[str]
     expected_remove: List[DaemonDescription]
 
 
-@pytest.mark.parametrize("service_type,placement,hosts,daemons,expected,expected_add,expected_remove",
+@pytest.mark.parametrize("service_type,placement,hosts,daemons,rank_map,post_rank_map,expected,expected_add,expected_remove",
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/tests/test_scheduling.py:389]: Bad feeling about that. This is going to make the test matrix explode.

File: src/pybind/mgr/cephadm/services/nfs.py [hunk 8]
```diff
@@ -89,24 +135,30 @@ def get_cephadm_config() -> Dict[str, Any]:
...
+        cmd = [
+            'rados',
+            '-n', f"mgr.{self.mgr.get_mgr_id()}",
+            '-k', str(self.mgr.get_ceph_option('keyring')),
+            '-p', cast(str, spec.pool),
+        ]
+        if spec.namespace:
+            cmd += ['--namespace', spec.namespace]
+        result = subprocess.run(
+            cmd + ['get', objname],
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/pybind/mgr/cephadm/services/nfs.py:148]: Are you sure?  It should succeed if the object exists (even if the object is 0 bytes)

File: src/pybind/mgr/cephadm/tests/test_scheduling.py (hunk 7 above)
Comment (liewegas) [reply src/pybind/mgr/cephadm/tests/test_scheduling.py:389]: for the non-ranked test cases, these are just None, None.  I thought about making a separate test but that would just duplicate the test code... or put it in a helper that just passed None, None for the older test cases.

Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/tests/test_scheduling.py:389]: I'm concerned that we essentially have to test

(different number of hosts) * (different existing placement on those hosts) * (different networks) * (different filter_new_host functions) * (allow_colo = yes and allow_colo = no) * (primary_daemon_type) * (per_host_daemon_type) * (different values for the rank_map). 

And that scares me.

**Review by varshar16 — CHANGES_REQUESTED** [review]
Please update the pool name in nfs tests here https://github.com/ceph/ceph/blob/master/qa/tasks/cephfs/test_nfs.py

**Comment by liewegas** [comment]
@varshar16 I went back to ``nfs-ganesha`` for the default pool name.. still undecided whether we should change this to ``.nfs-ganesha`` or ``.nfs``

File: src/pybind/mgr/cephadm/services/nfs.py [hunk 9]
```diff
@@ -141,7 +173,7 @@ def remove_rgw_keyring(self, daemon: DaemonDescription) -> None:
         entity: AuthEntity = self.get_auth_entity(f'{daemon_id}-rgw')
 
         logger.info(f'Removing key for {entity}')
-        ret, out, err = self.mgr.check_mon_command({
+        ret, out, err = self.mgr.mon_command({
```
Comment (varshar16) [inline src/pybind/mgr/cephadm/services/nfs.py:251]: Let it be `check_mon_command`, it checks the return value. We can remove `ret, out, err`.

File: src/pybind/mgr/cephadm/services/nfs.py [hunk 10]
```diff
@@ -82,12 +82,16 @@ def generate_config(self, daemon_spec: CephadmDaemonDeploySpec) -> Tuple[Dict[st
...
-                           rgw_user=rgw_user,
-                           url=spec.rados_config_location())
+            context = {
+                "user": rados_user,
+                "nodeid": f'{daemon_spec.service_name}.{daemon_spec.rank}',
+                "pool": spec.pool,
+                "namespace": spec.namespace if spec.namespace else '',
+                "rgw_user": rgw_user,
+                "url": spec.rados_config_location(),
+                "port": daemon_spec.ports[0] if daemon_spec.ports else 2049,
```
Comment (varshar16) [inline src/pybind/mgr/cephadm/services/nfs.py:108]: Add a note saying `2049` is the default port used by ganesha or define constant for it `DefaultGaneshaPort=2049`

File: src/pybind/mgr/cephadm/services/nfs.py [hunk 11]
```diff
@@ -171,6 +183,53 @@ def create_rgw_keyring(self, daemon_spec: CephadmDaemonDeploySpec) -> str:
...
+            result = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+                                    timeout=10)
+            if result.returncode:
+                self.mgr.log.warning(
+                    f'ganesha-rados-grace tool failed: {result.stderr.decode("utf-8")}'
+                )
+                raise RuntimeError(f'grace tool failed: {result.stderr.decode("utf-8")}')
+
+        finally:
+            ret, out, err = self.mgr.mon_command({
```
Comment (varshar16) [inline src/pybind/mgr/cephadm/services/nfs.py:240]: use `check_mon_command` instead of `mon_command`

File: src/pybind/mgr/orchestrator/module.py [hunk 12]
```diff
@@ -1124,6 +1125,7 @@ def _apply_nfs(self,
             service_id=svc_id,
             pool=pool,
             namespace=namespace,
+            port=port,
```
Comment (varshar16) [inline src/pybind/mgr/orchestrator/module.py:1128]: Update doc about it

File: src/pybind/mgr/cephadm/services/nfs.py [hunk 13]
```diff
@@ -57,7 +57,12 @@ def fence_old_ranks(self,
     def config(self, spec: NFSServiceSpec, daemon_id: str) -> None:  # type: ignore
         assert self.TYPE == spec.service_type
         assert spec.pool
-        self.mgr._check_pool_exists(spec.pool, spec.service_name())
+        pool_list = [p['pool_name'] for p in self.mgr.get_osdmap().dump().get('pools', [])]
+
+        if spec.pool not in pool_list:
+            self.mgr.check_mon_command({'prefix': 'osd pool create', 'pool': spec.pool})
+            self.mgr.check_mon_command({'prefix': 'osd pool application enable',
+                                        'pool': spec.pool, 'app': 'nfs'})
```
Comment (varshar16) [inline src/pybind/mgr/cephadm/services/nfs.py:65]: Move this to orchestrator code instead of being specific to cephadm ?

Comment (varshar16) [reply src/pybind/mgr/cephadm/services/nfs.py:65]: If we plan to use `nfs-ganesha` as default pool. Then we can remove redundant code from nfs module too
https://github.com/ceph/ceph/blob/master/src/pybind/mgr/nfs/cluster.py#L68-L73

File: src/python-common/ceph/deployment/service_spec.py [hunk 14]
```diff
@@ -679,10 +679,10 @@ def __init__(self,
             config=config, networks=networks)
 
         #: RADOS pool where NFS client recovery data is stored.
-        self.pool = pool
+        self.pool = pool or 'nfs-ganesha'
```
Comment (varshar16) [inline src/python-common/ceph/deployment/service_spec.py:682]: Define a constant for default pool name. Then we can remove from nfs module https://github.com/ceph/ceph/blob/master/src/pybind/mgr/nfs/utils.py#L3

File: doc/cephfs/fs-nfs-exports.rst [hunk 15]
```diff
@@ -67,7 +67,7 @@ Delete NFS Ganesha Cluster
 
 .. code:: bash
 
-    $ ceph nfs cluster delete <clusterid>
+    $ ceph nfs cluster rm <clusterid>
```
Comment (varshar16) [inline doc/cephfs/fs-nfs-exports.rst:70]: Why is this change required?

File: src/pybind/mgr/nfs/cluster.py [hunk 16]
```diff
@@ -40,12 +40,31 @@ def _get_common_conf_obj_name(self):
...
-        completion = self.mgr.apply_nfs(spec)
-        orchestrator.raise_if_exception(completion)
+    def _call_orch_apply_nfs(self, placement, virtual_ip):
+        if virtual_ip:
+            # nfs + ingress
+            # run NFS on non-standard port
+            spec = NFSServiceSpec(service_type='nfs', service_id=self.cluster_id,
+                                  pool=self.pool_name, namespace=self.pool_ns,
+                                  placement=PlacementSpec.from_string(placement),
+                                  port=12049)
```
Comment (varshar16) [inline src/pybind/mgr/nfs/cluster.py:52]: Please define constant for `12049` or add some comment about it.

File: src/pybind/mgr/nfs/cluster.py [hunk 17]
```diff
@@ -40,12 +40,31 @@ def _get_common_conf_obj_name(self):
     def _get_user_conf_obj_name(self):
         return f'userconf-nfs.{self.cluster_id}'
 
-    def _call_orch_apply_nfs(self, placement):
-        spec = NFSServiceSpec(service_type='nfs', service_id=self.cluster_id,
-                              pool=self.pool_name, namespace=self.pool_ns,
-                              placement=PlacementSpec.from_string(placement))
-        completion = self.mgr.apply_nfs(spec)
-        orchestrator.raise_if_exception(completion)
+    def _call_orch_apply_nfs(self, placement, virtual_ip):
```
Comment (varshar16) [inline src/pybind/mgr/nfs/cluster.py:45]: `def _call_orch_apply_nfs(self, placement, virtual_ip=None):`

File: src/pybind/mgr/nfs/cluster.py [hunk 18]
```diff
@@ -147,16 +152,24 @@ def _show_nfs_cluster_info(self, cluster_id):
...
                     continue
-        return host_ip
+
+        sc = self.mgr.describe_service(service_type='ingress')
+        orchestrator.raise_if_exception(sc)
+        for i in sc.result:
+            spec = cast(IngressSpec, i.spec)
+            if spec.backend_service == f'nfs.{cluster_id}':
+                r['virtual_ip'] = i.virtual_ip.split('/')[0]
+                r['port'] = i.ports[0] or 2049
```
Comment (varshar16) [inline src/pybind/mgr/nfs/cluster.py:169]: Please keep the comment about 2049 ganesha port. `# Default ganesha port`

**Comment by varshar16** [comment]
> @varshar16 I went back to `nfs-ganesha` for the default pool name.. still undecided whether we should change this to `.nfs-ganesha` or `.nfs`

If we are planning to use `nfs-ganesha` as default pool at orchestrator level. Then I prefer changing it.

**Review by varshar16 — CHANGES_REQUESTED** [review]
This particular [commit](https://github.com/ceph/ceph/pull/41007/commits/fd21c3c66300b7baa4e281ff7720258be0311263) will break the tests and cause backward compatibility issues.

File: doc/cephfs/fs-nfs-exports.rst (hunk 15 above)
Comment (liewegas) [reply doc/cephfs/fs-nfs-exports.rst:70]: it just makes the CLI consistent; everywhere else we use ``rm`` instead of ``remove`` or ``delete``.  are there any users of this besides the tests?

**Review by mgfritch — APPROVED** [review]
lgtm! manually tested a few failover scenarios that worked as expected!

File: PendingReleaseNotes [hunk 19]
```diff
@@ -22,6 +22,20 @@
   support an NFS export of both ``rgw`` and ``cephfs`` from a single
   NFS cluster instance.
 
+* The ``nfs cluster update`` command has been removed.  You can modify
+  the placement of an existing NFS service (and/or its associated
+  ingress service) using ``orch ls --export`` and ``orch apply -i
+  ...``.
```
Comment (liewegas) [inline PendingReleaseNotes:28]: @sebastian-philipp @varshar16 We need to make a call here on whether to (1) remove this, (2) keep it but require that all relevant arguments be provided (i.e., wrap ``orch apply nfs`` and do not support incremental changes/patches), or (3) implement a more complex version that allows incremental changes.

I lean toward (1) since it's the simplest and this is convenience wrapper only, and because we can always come back later with something more friendly if we see a need.

**Review by mgfritch — APPROVED** [review]
additional changes ltgm

File: PendingReleaseNotes (hunk 19 above)
Comment (varshar16) [reply PendingReleaseNotes:28]: How about making it a wrap around `orch apply -i nfs.yaml`  instead of `orch apply nfs` ? We would need additional checks to ensure the yaml is related to `nfs` service only.

**Review by varshar16 — CHANGES_REQUESTED** [review]
In the following commands `clusterid` requirement is made optional which breaks cli.
```
nfs cluster config reset [<clusterid>]                                                                                 Reset NFS-Ganesha Config to default
nfs cluster create [<clusterid>] [<placement>] [--ingress] [<virtual_ip>]                                              Create an NFS Cluster
nfs cluster delete [<clusterid>]                                                                                       Removes an NFS Cluster (DEPRECATED)
nfs cluster rm [<clusterid>]                                                                                           Removes an NFS Cluster
nfs export ls [<clusterid>] [--detailed]                                                                               List exports of a NFS cluster
```
`binding` is required in the following commands
```
nfs export create cephfs <fsname> <clusterid> [<binding>] [--readonly] [<path>]                                        Create a cephfs export
nfs export delete <clusterid> [<binding>]                                                                              Delete a cephfs export (DEPRECATED)
nfs export get <clusterid> [<binding>]                                                                                 Fetch a export of a NFS cluster given the pseudo path/binding
nfs export rm <clusterid> [<binding>]                                                                                  Remove a cephfs export
```

File: doc/cephadm/nfs.rst [hunk 20]
```diff
@@ -1,32 +1,35 @@
...
+Cephadm deploys NFS Ganesha daemon (or set of daemons).  The configuration for
+NFS is stored in the ``nfs-ganesha`` pool and exports are managed via the
+``ceph nfs export ...`` commands and via the dashboard.
 
 To deploy a NFS Ganesha gateway, run the following command:
 
 .. prompt:: bash #
 
-    ceph orch apply nfs *<svc_id>* *<pool>* *<namespace>* --placement="*<num-daemons>* [*<host1>* ...]"
+    ceph orch apply nfs *<svc_id>* [--port *<port>*] [--placement ...]
```
Comment (varshar16) [inline doc/cephadm/nfs.rst:25]: Add a link to [placement specification section](https://docs.ceph.com/en/latest/cephadm/service-management/#orchestrator-cli-placement-spec)

File: src/cephadm/cephadm [hunk 21]
```diff
@@ -4329,7 +4293,8 @@ def command_deploy(ctx):
 
     elif daemon_type == NFSGanesha.daemon_type:
         if not ctx.reconfig and not redeploy:
```
Comment (varshar16) [inline src/cephadm/cephadm:4295]:
```suggestion
        if not ctx.reconfig and not redeploy and not daemon_ports:
```

File: src/pybind/mgr/nfs/cluster.py [hunk 22]
```diff
@@ -126,16 +144,29 @@ def _show_nfs_cluster_info(self, cluster_id):
                 ('2404:6800:4009:80d::200e', 2049, 0, 0))]
```
Comment (varshar16) [inline src/pybind/mgr/nfs/cluster.py:144]: Remove this comment

File: src/pybind/mgr/nfs/cluster.py [hunk 23]
```diff
@@ -126,16 +144,29 @@ def _show_nfs_cluster_info(self, cluster_id):
...
+                    backends.append({
                             "hostname": cluster.hostname,
-                            "ip": list(set([ip[4][0] for ip in socket.getaddrinfo(
-                                cluster.hostname, 2049, flags=socket.AI_CANONNAME,
-                                type=socket.SOCK_STREAM)])),
-                            "port": 2049  # Default ganesha port
+                            "ip": cluster.ip or resolve_ip(cluster.hostname),
+                            "port": cluster.ports[0]
                             })
                 except socket.gaierror:
```
Comment (varshar16) [inline src/pybind/mgr/nfs/cluster.py:152]: Change this exception to `OrchestratorError`

https://github.com/ceph/ceph/blob/f8e872ce2fc7d6f0f960d51942e3e9cc4d8cc265/src/pybind/mgr/cephadm/utils.py#L107-L117

**Comment by varshar16** [comment]
> In the following commands `clusterid` requirement is made optional which breaks cli.
> 
> ```
> nfs cluster config reset [<clusterid>]                                                                                 Reset NFS-Ganesha Config to default
> nfs cluster create [<clusterid>] [<placement>] [--ingress] [<virtual_ip>]                                              Create an NFS Cluster
> nfs cluster delete [<clusterid>]                                                                                       Removes an NFS Cluster (DEPRECATED)
> nfs cluster rm [<clusterid>]                                                                                           Removes an NFS Cluster
> nfs export ls [<clusterid>] [--detailed]                                                                               List exports of a NFS cluster
> ```
> 
> `binding` is required in the following commands
> 
> ```
> nfs export create cephfs <fsname> <clusterid> [<binding>] [--readonly] [<path>]                                        Create a cephfs export
> nfs export delete <clusterid> [<binding>]                                                                              Delete a cephfs export (DEPRECATED)
> nfs export get <clusterid> [<binding>]                                                                                 Fetch a export of a NFS cluster given the pseudo path/binding
> nfs export rm <clusterid> [<binding>]                                                                                  Remove a cephfs export
> ```

Sorry, it is not due to this PR. I have fixed this issue here https://github.com/ceph/ceph/pull/41314

File: src/pybind/mgr/cephadm/services/nfs.py [hunk 24]
```diff
@@ -55,9 +55,11 @@ def fence_old_ranks(self,
                         self.mgr.spec_store.save_rank_map(spec.service_name(), rank_map)
 
     def config(self, spec: NFSServiceSpec, daemon_id: str) -> None:  # type: ignore
+        from nfs.cluster import create_ganesha_pool
+
         assert self.TYPE == spec.service_type
         assert spec.pool
-        self.mgr._check_pool_exists(spec.pool, spec.service_name())
+        create_ganesha_pool(self.mgr, spec.pool)
```
Comment (varshar16) [inline src/pybind/mgr/cephadm/services/nfs.py:62]: Let's move pool creation to `orch` interface. Then we need not create it in `nfs`, `cephadm` or `rook`.

File: doc/cephfs/fs-nfs-exports.rst (hunk 15 above)
Comment (varshar16) [reply doc/cephfs/fs-nfs-exports.rst:70]: @liewegas can you please update the commit message with the reason for removal ?

**Comment by liewegas** [comment]
> In the following commands `clusterid` requirement is made optional which breaks cli.
> 
> ```
> nfs cluster config reset [<clusterid>]                                                                                 Reset NFS-Ganesha Config to default
> nfs cluster create [<clusterid>] [<placement>] [--ingress] [<virtual_ip>]                                              Create an NFS Cluster
> nfs cluster delete [<clusterid>]                                                                                       Removes an NFS Cluster (DEPRECATED)
> nfs cluster rm [<clusterid>]                                                                                           Removes an NFS Cluster
> nfs export ls [<clusterid>] [--detailed]                                                                               List exports of a NFS cluster
> ```
> 
> `binding` is required in the following commands
> 
> ```
> nfs export create cephfs <fsname> <clusterid> [<binding>] [--readonly] [<path>]                                        Create a cephfs export
> nfs export delete <clusterid> [<binding>]                                                                              Delete a cephfs export (DEPRECATED)
> nfs export get <clusterid> [<binding>]                                                                                 Fetch a export of a NFS cluster given the pseudo path/binding
> nfs export rm <clusterid> [<binding>]                                                                                  Remove a cephfs export
> ```

I don't think I changed clusterid at all anywhere...?

**Comment by sebastian-philipp** [comment]
@Daniel-Pivonka didn't found anything bad :+1:

File: PendingReleaseNotes (hunk 19 above)
Comment (sebastian-philipp) [reply PendingReleaseNotes:34]: The calls to `orch apply -i nfs.yaml` and `orch apply nfs` are identical. I'd be ok with (1)

File: src/pybind/mgr/cephadm/services/nfs.py (hunk 24 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/services/nfs.py:62]: that would require breaking changes in rook, right? In any case what we need is idempotent.

**Comment by varshar16** [comment]
> > In the following commands `clusterid` requirement is made optional which breaks cli.
> > ```
> > nfs cluster config reset [<clusterid>]                                                                                 Reset NFS-Ganesha Config to default
> > nfs cluster create [<clusterid>] [<placement>] [--ingress] [<virtual_ip>]                                              Create an NFS Cluster
> > nfs cluster delete [<clusterid>]                                                                                       Removes an NFS Cluster (DEPRECATED)
> > nfs cluster rm [<clusterid>]                                                                                           Removes an NFS Cluster
> > nfs export ls [<clusterid>] [--detailed]                                                                               List exports of a NFS cluster
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > `binding` is required in the following commands
> > ```
> > nfs export create cephfs <fsname> <clusterid> [<binding>] [--readonly] [<path>]                                        Create a cephfs export
> > nfs export delete <clusterid> [<binding>]                                                                              Delete a cephfs export (DEPRECATED)
> > nfs export get <clusterid> [<binding>]                                                                                 Fetch a export of a NFS cluster given the pseudo path/binding
> > nfs export rm <clusterid> [<binding>]                                                                                  Remove a cephfs export
> > ```
> 
> I don't think I changed clusterid at all anywhere...?

Sorry, the issue was not due to this PR. It is fixed by https://github.com/ceph/ceph/pull/41314

File: src/pybind/mgr/nfs/cluster.py [hunk 25]
```diff
@@ -89,7 +89,15 @@ def delete_config_obj(self):
...
-    def create_nfs_cluster(self, cluster_id, placement, virtual_ip):
+    def create_nfs_cluster(self,
+                           cluster_id: str,
+                           placement: Optional[str],
+                           virtual_ip: Optional[str],
+                           ingress: Optional[bool] = None):
+        if virtual_ip and not ingress:
+            raise orchestrator.OrchestratorError('virtual_ip can only be provided with ingress enabled')
+        if not virtual_ip and ingress:
+            raise orchestrator.OrchestratorError('ingress currently requires a virtual_ip')
```
Comment (varshar16) [inline src/pybind/mgr/nfs/cluster.py:100]: Let this code be within try block and raise `NFSInvalidOperation` exception instead of `OrchestratorError`

File: src/pybind/mgr/nfs/module.py [hunk 26]
```diff
@@ -70,14 +70,11 @@ def _cmd_nfs_cluster_create(self,
...
-        if ingress and not virtual_ip:
-            return (-errno.EINVAL, '',
-                    '--ingress current requires --virtual-ip')
-        return self.nfs.create_nfs_cluster(cluster_id=clusterid, placement=placement,
-                                           virtual_ip=virtual_ip)
+        try:
+            return self.nfs.create_nfs_cluster(cluster_id=clusterid, placement=placement,
+                                               virtual_ip=virtual_ip, ingress=ingress)
+        except orchestrator.OrchestratorError as e:
+            return (-errno.EINVAL, '', str(e))
```
Comment (varshar16) [inline src/pybind/mgr/nfs/module.py:77]: Here `try-except` block is not required, if above suggested changes are made

Comment (varshar16) [reply src/pybind/mgr/nfs/module.py:77]: Please update doc about virtual ip and ingress https://ceph--41007.org.readthedocs.build/en/41007/cephfs/fs-nfs-exports/

**Review by varshar16 — CHANGES_REQUESTED** [review]
If we are going ahead with [option 1](https://github.com/ceph/ceph/pull/41007#discussion_r628279921) then please update the [doc](https://ceph--41007.org.readthedocs.build/en/41007/cephfs/fs-nfs-exports/#update-nfs-ganesha-cluster) and create a tracker ticket for the new `update` interface.

**Comment by liewegas** [comment]
> If we are going ahead with [option 1](https://github.com/ceph/ceph/pull/41007#discussion_r628279921) then please update the [doc](https://ceph--41007.org.readthedocs.build/en/41007/cephfs/fs-nfs-exports/#update-nfs-ganesha-cluster) and create a tracker ticket for the new `update` interface.

https://tracker.ceph.com/issues/50972
https://github.com/ceph/ceph/pull/41534

**Review by varshar16 — APPROVED** [review]
Thanks @liewegas for addressing my comments.
